src/components/implementation/no_interface/llbooter/boot_deps.h

```diff
@@ -135,229 +189,391 @@ boot_newcomp_sinv_alloc(spdid_t spdid)
-	const char prefix[] = "shmem";
+	struct cos_compinfo  *boot_info = boot_spd_compinfo_get(0);
+	struct comp_cap_info *acomp     = &new_comp_cap_info[srcid];
+	struct cos_compinfo  *resci     = boot_spd_compinfo_get(dstid);
+	struct cos_aep_info  *initaep   = boot_spd_initaep_get(srcid);
+	int ret = 0;
 
-	ret = !strncmp(comp_name, prefix, sizeof(prefix)-1);
+	if (srcid > num_cobj || dstid > num_cobj) return -1;
+
+	if (initaep->thd) {
+		thdcap_t t;
+
+		ret = cos_cap_cpy_at(resci, thdslot, boot_info, initaep->thd);
+		assert(ret == 0);
+
+		if (acomp->is_sched) {
+			assert(initaep->rcv && initaep->tc);
+			ret = cos_cap_cpy_at(resci, rcvslot, boot_info, initaep->rcv);
+			assert(ret == 0);
+
+			ret = cos_cap_cpy_at(resci, tcslot, boot_info, initaep->tc);
+			assert(ret == 0);
+		}
+
+		return 0;
+	}
 
-	return ret;
+	return -1;
 }
 
-void
-boot_thd_done(void)
+static int 
+boot_comp_info_get(spdid_t dstid, spdid_t srcid, pgtblcap_t pgtslot, captblcap_t captslot, compcap_t ccslot, spdid_t *psid)
 {
-	printc("Done booting\n\n");
+	struct cos_compinfo *boot_info = boot_spd_compinfo_get(0);
+	struct cos_compinfo *a_ci, *resci;
+	int ret = 0;
+
+	if (srcid > num_cobj || dstid > num_cobj) return -1;
+
+	a_ci  = boot_spd_compinfo_get(srcid);
+	resci = boot_spd_compinfo_get(dstid);
+	assert(a_ci);
+	assert(resci);
+
+	ret = cos_cap_cpy_at(resci, pgtslot, boot_info, a_ci->pgtbl_cap);
+	assert(ret == 0);
+	ret = cos_cap_cpy_at(resci, captslot, boot_info, a_ci->captbl_cap);
+	assert(ret == 0);
+	ret = cos_cap_cpy_at(resci, ccslot, boot_info, a_ci->comp_cap);
+	assert(ret == 0);
+	*psid = new_comp_cap_info[srcid].parent_spdid;
+
+	return 0;
+}
 
-	global_idx++;
-	if (global_idx >= INIT_THDS_SIZE) goto done;
+static int 
+boot_comp_info_iter(spdid_t dstid, pgtblcap_t pgtslot, captblcap_t captslot, compcap_t ccslot, spdid_t *srcid, spdid_t *srcpsid)
+{
+	static int iter_idx = 0; /* including llbooter! */
+	int remain = 0, ret = 0;
 
-	while (init_thds[global_idx] == 0) global_idx++;
-	if (global_idx >= INIT_THDS_SIZE) goto done;
-	cos_thd_switch(init_thds[global_idx]);
+	if (iter_idx > num_cobj) return -1;
 
-	/*
-	 * ERROR, we should not end up back in this thread,
-	 * booter runs us just once to initialize
-	 */
-	assert(0);
-
-done:
-	/* There are not more threads to run */
-	printc("\nSystem done\n");
-	while(1);
+	*srcid = iter_idx;
+	iter_idx ++;
+	/*remaining */
+	remain = num_cobj + 1 - iter_idx;
+	ret = boot_comp_info_get(dstid, *srcid, pgtslot, captslot, ccslot, srcpsid);
+	if (ret) return -1;
 
+	return remain;
 }
 
-long
-boot_pgtbl_cap_transfer(int dst, int src, int cap_slot)
+static int
+boot_comp_frontier_get(spdid_t dstid, spdid_t srcid, vaddr_t *vasfr, capid_t *capfr)
 {
-	return cos_cap_cpy_at(new_comp_cap_info[dst].compinfo, cap_slot, boot_info(),
-			new_comp_cap_info[src].compinfo->pgtbl_cap);
+	struct cos_compinfo *a_ci = boot_spd_compinfo_get(srcid);
+
+	assert(a_ci);
+	*vasfr = a_ci->vas_frontier;
+	*capfr = a_ci->cap_frontier;
```

Comment:
if we had the time to do defensive programming here, we'd disable the ability to do allocations within this component using these frontiers.  We are going from a mode of "allocate the resources for the component" to "you manage your own namespaces", and it would be good if we had checks to ensure that transition.

I see strange bugs for people in the future...and we don't have the time now to prevent them.

Reply:
This is used by resource manager or a component to get the frontiers of components from booter after the boot phase. Once they get this from booter, they're managing their own namespace. This is the only time they need to get some info from the booter because booter would have bumped the frontiers in the boot phase.